### DISCUSSION TITLE
Help to Build a Plugin for Unreal 5.3 and Unreal 5.4

### DISCUSSION THREAD

**Discussion by NISP2709** [JSBSim-Team/jsbsim#1180] [Q&A]
When I am trying to open the project the Unreal Projects states that to build the plugin manually, So I am writing the command to build it for both the engine version. For the Both the Engine I getting different errors. This are the following commands and errors that I am getting while building the plugin:

_I am not much of a programmer, Please guide if I am making some mistakes._

**Command Error for Unreal 5.3:**
```
RunUAT.bat BuildPlugin -plugin="C:\Users\DigiCon\Downloads\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\JSBSimFlightDynamicsModel.uplugin" -package="D:\Nish\Build\5.3\JSBSimFlightDynamicsModel"
Running AutomationTool...
Using bundled DotNet SDK version: 6.0.302
Starting AutomationTool...
Parsing command line: BuildPlugin -plugin=C:\Users\DigiCon\Downloads\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\JSBSimFlightDynamicsModel.uplugin -package=D:\Nish\Build\5.3\JSBSimFlightDynamicsModel
Initializing script modules...
Total script module initialization time: 0.63 s.
Executing commands...
Copying 573 file(s) using max 64 thread(s)
Reading plugin from D:\Nish\Build\5.3\JSBSimFlightDynamicsModel\HostProject\Plugins\JSBSimFlightDynamicsModel\JSBSimFlightDynamicsModel.uplugin...
Building plugin for host platforms: Win64
Running: C:\Program Files\Epic Games\UE_5.3\Engine\Binaries\ThirdParty\DotNet\6.0.302\windows\dotnet.exe "C:\Program Files\Epic Games\UE_5.3\Engine\Binaries\DotNET\UnrealBuildTool\UnrealBuildTool.dll" UnrealEditor Win64 Development -Project="D:\Nish\Build\5.3\JSBSimFlightDynamicsModel\HostProject\HostProject.uproject" -plugin="D:\Nish\Build\5.3\JSBSimFlightDynamicsModel\HostProject\Plugins\JSBSimFlightDynamicsModel\JSBSimFlightDynamicsModel.uplugin" -noubtmakefiles -manifest="D:\Nish\Build\5.3\JSBSimFlightDynamicsModel\HostProject\Saved\Manifest-UnrealEditor-Win64-Development.xml" -nohotreload -log="C:\Users\DigiCon\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_5.3\UBT-UnrealEditor-Win64-Development.txt"
Log file: C:\Users\DigiCon\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_5.3\UBT-UnrealEditor-Win64-Development.txt
JSBSim Path Check OK: D:\Nish\Build\5.3\JSBSimFlightDynamicsModel\HostProject\Plugins\JSBSimFlightDynamicsModel\Source\ThirdParty\JSBSim\Lib\JSBSim.dll
Total execution time: 8.89 seconds
Unhandled exception: System.ArgumentNullException: Value cannot be null. (Parameter 'element')
   at System.Attribute.GetCustomAttributes(MemberInfo element, Type attributeType, Boolean inherit)
   at System.Reflection.CustomAttributeExtensions.GetCustomAttributes[T](MemberInfo element)
   at UnrealBuildTool.ModuleRules.IsValidForTarget(Type ModuleType, ReadOnlyTargetRules TargetRules, String& InvalidReason) in D:\build\++UE5\Sync\Engine\Saved\CsTools\Engine\Source\Programs\UnrealBuildTool\Configuration\ModuleRules.cs:line 1623
   at UnrealBuildTool.UEBuildTarget.AddAllValidModulesToTarget(ILogger Logger) in D:\build\++UE5\Sync\Engine\Saved\CsTools\Engine\Source\Programs\UnrealBuildTool\Configuration\UEBuildTarget.cs:line 3886
   at UnrealBuildTool.UEBuildTarget.PreBuildSetup(ILogger Logger) in D:\build\++UE5\Sync\Engine\Saved\CsTools\Engine\Source\Programs\UnrealBuildTool\Configuration\UEBuildTarget.cs:line 3354
   at UnrealBuildTool.UEBuildTarget.Create(TargetDescriptor Descriptor, Boolean bSkipRulesCompile, Boolean bForceRulesCompile, Boolean bUsePrecompiled, UnrealIntermediateEnvironment IntermediateEnvironment, ILogger Logger) in D:\build\++UE5\Sync\Engine\Saved\CsTools\Engine\Source\Programs\UnrealBuildTool\Configuration\UEBuildTarget.cs:line 1366
   at UnrealBuildTool.BuildMode.CreateMakefileAsync(BuildConfiguration BuildConfiguration, TargetDescriptor TargetDescriptor, ISourceFileWorkingSet WorkingSet, ILogger Logger) in D:\build\++UE5\Sync\Engine\Saved\CsTools\Engine\Source\Programs\UnrealBuildTool\Modes\BuildMode.cs:line 1114
   at UnrealBuildTool.BuildMode.BuildAsync(List`1 TargetDescriptors, BuildConfiguration BuildConfiguration, ISourceFileWorkingSet WorkingSet, BuildOptions Options, FileReference WriteOutdatedActionsFile, ILogger Logger, Boolean bSkipPreBuildTargets) in D:\build\++UE5\Sync\Engine\Saved\CsTools\Engine\Source\Programs\UnrealBuildTool\Modes\BuildMode.cs:line 396
   at UnrealBuildTool.BuildMode.ExecuteAsync(CommandLineArguments Arguments, ILogger Logger) in D:\build\++UE5\Sync\Engine\Saved\CsTools\Engine\Source\Programs\UnrealBuildTool\Modes\BuildMode.cs:line 252
   at UnrealBuildTool.UnrealBuildTool.Main(String[] ArgumentsArray) in D:\build\++UE5\Sync\Engine\Saved\CsTools\Engine\Source\Programs\UnrealBuildTool\UnrealBuildTool.cs:line 659
Took 9.24s to run dotnet.exe, ExitCode=6
UnrealBuildTool failed. See log for more details. (C:\Users\DigiCon\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_5.3\UBT-UnrealEditor-Win64-Development.txt)
AutomationTool executed for 0h 0m 11s
AutomationTool exiting with ExitCode=6 (6)
BUILD FAILED
```

**Command Error for Unreal 5.4:**
```
RunUAT.bat BuildPlugin -plugin="C:\Users\DigiCon\Downloads\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\JSBSimFlightDynamicsModel.uplugin" -package="D:\Nish\Build\5.4\JSBSimFlightDynamicsModel"
Running AutomationTool...
Using bundled DotNet SDK version: 6.0.302
Starting AutomationTool...
Parsing command line: BuildPlugin -plugin=C:\Users\DigiCon\Downloads\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\JSBSimFlightDynamicsModel.uplugin -package=D:\Nish\Build\5.4\JSBSimFlightDynamicsModel
Initializing script modules...
Total script module initialization time: 0.63 s.
Using C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\MSBuild.exe
Executing commands...
Copying 573 file(s) using max 64 thread(s)
Reading plugin from D:\Nish\Build\5.4\JSBSimFlightDynamicsModel\HostProject\Plugins\JSBSimFlightDynamicsModel\JSBSimFlightDynamicsModel.uplugin...
Building plugin for host platforms: Win64
Running: C:\Program Files\Epic Games\UE_5.4\Engine\Binaries\ThirdParty\DotNet\6.0.302\windows\dotnet.exe "C:\Program Files\Epic Games\UE_5.4\Engine\Binaries\DotNET\UnrealBuildTool\UnrealBuildTool.dll" UnrealEditor Win64 Development -Project="D:\Nish\Build\5.4\JSBSimFlightDynamicsModel\HostProject\HostProject.uproject" -plugin="D:\Nish\Build\5.4\JSBSimFlightDynamicsModel\HostProject\Plugins\JSBSimFlightDynamicsModel\JSBSimFlightDynamicsModel.uplugin" -noubtmakefiles -manifest="D:\Nish\Build\5.4\JSBSimFlightDynamicsModel\HostProject\Saved\Manifest-UnrealEditor-Win64-Development.xml" -nohotreload -log="C:\Users\DigiCon\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_5.4\UBA-UnrealEditor-Win64-Development.txt"
Log file: C:\Users\DigiCon\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_5.4\UBA-UnrealEditor-Win64-Development.txt
JSBSim Path Check OK: D:\Nish\Build\5.4\JSBSimFlightDynamicsModel\HostProject\Plugins\JSBSimFlightDynamicsModel\Source\ThirdParty\JSBSim\Lib\JSBSim.dll
Parsing headers for UnrealEditor
  Running Internal UnrealHeaderTool D:\Nish\Build\5.4\JSBSimFlightDynamicsModel\HostProject\HostProject.uproject D:\Nish\Build\5.4\JSBSimFlightDynamicsModel\HostProject\Intermediate\Build\Win64\UnrealEditor\Development\UnrealEditor.uhtmanifest -WarningsAsErrors -installed
Total of 0 written
C:\Program Files\Epic Games\UE_5.4\Engine\Plugins\Experimental\UserToolBoxBasicCommand\Source\UserToolBoxBasicCommand\Public\EngineCommand.h(12): Error: Class 'UEngineCommand' shares engine name 'EngineCommand' with struct 'FEngineCommand' in D:\Nish\Build\5.4\JSBSimFlightDynamicsModel\HostProject\Plugins\JSBSimFlightDynamicsModel\Source\JSBSimFlightDynamicsModel\Public\FDMTypes.h(140)
Total execution time: 72.46 seconds
Took 72.69s to run dotnet.exe, ExitCode=6
UnrealBuildTool failed. See log for more details. (C:\Users\DigiCon\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_5.4\UBA-UnrealEditor-Win64-Development.txt)
AutomationTool executed for 0h 1m 14s
AutomationTool exiting with ExitCode=6 (6)
BUILD FAILED
```

### Please Let m know where I am making mistake

**Comment by bcoconni** [comment]
It has been mentioned in #1051 that the JSBSim plugin is compatible with UE 5.0-5.3. Note sure about 5.4. @gallonmate, any idea ?

**Reply by bcoconni** [reply]
@NISP2709, it is mentioned in the error report that you have posted that more details are available in logs:
`UnrealBuildTool failed. See log for more details. (C:\Users\DigiCon\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_5.3\UBT-UnrealEditor-Win64-Development.txt)`
and
`UnrealBuildTool failed. See log for more details. (C:\Users\DigiCon\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_5.4\UBA-UnrealEditor-Win64-Development.txt)`

Would you mind posting the content of these logs ?

**Reply by NISP2709** [reply]
Here are the log files for both the version that is generated.

Thanks.
[UBA-UnrealEditor-Win64-Development_5.4.txt](https://github.com/user-attachments/files/17600054/UBA-UnrealEditor-Win64-Development_5.4.txt)
[UBT-UnrealEditor-Win64-Development_5.3.txt](https://github.com/user-attachments/files/17600064/UBT-UnrealEditor-Win64-Development_5.3.txt)

**Comment by gallonmate** [comment]
Did you already follow all the steps in the Read Me file? https://github.com/JSBSim-Team/jsbsim/blob/master/UnrealEngine/README.md 

Basically:
First open JSBSimForUnreal.sln and build it in Visual Studio.
Then you can right click on UnrealEngine\UEReferenceApp.uproject and select "Generate Visual Studio project files". After some time the file UEReferenceApp.sln will appear beside the uproject file. Open it and build it in Visual Studio. Try these steps with UE5.3 first.

**Comment by tbuckup** [comment]
I confirm it cant build in 5.5.
Two things need to be fixed:
- First the toolchain needs to be the 14.38.33130, since 14.29 is not supported anymore. Instructions in `https://github.com/JSBSim-Team/jsbsim/tree/master/UnrealEngine` are not correct anymore.

- Second, one needs to edit the  `UEReferenceAppEditor.Target.cs` as :

```
 DefaultBuildSettings = BuildSettingsVersion.V5;
 CppStandard = CppStandardVersion.Cpp20;
 IncludeOrderVersion = EngineIncludeOrderVersion.Unreal5_5;
 BuildEnvironment = TargetBuildEnvironment.Shared;
```

**Reply by gallonmate** [reply]
Thanks. I also finally got around to testing UE5.4 and can confirm it's a similar issue. I will submit a pull request for these changes and to update the instructions found in the Readme file.

**Reply by adl67** [reply]
Thank you a lot! It works without any modify in UE5.4 but it don't works with 5.5 for me, in UE5.5 I have the error "exit from comand ""C:\Program Files\Epic Games\UE_5.5\Engine\Build\BatchFiles\Build.bat" UEReferenceAppEditor Win64 Development -Project="E:\Lavori\Unreal\jsbsim5.5\UnrealEngine\UEReferenceApp.uproject" -WaitMutex -FromMsBuild -architecture=x64" with code 6."

**Reply by gallonmate** [reply]
@adl67 can you post more of the log file or check it for other errors/messages? One common thing you can do is run the "CleanProject.bat", found in your project "jsbsim\UnrealEngine" folder, which will cleanup the build files, then you can generate VS solution, and try building the plugin.

**Reply by adl67** [reply]
Thank you Gallonmate, but I have the same error for UE 5.5 after clean.bat too with or without any modifies in the Source, this is the log and thank you for any help... 


Compilazione avviata alle 17:40...
1>------ Inizio compilazione: Progetto: EpicGames.UBA, Configurazione: Development Any CPU ------
2>------ Inizio compilazione: Progetto: EpicGames.Oodle, Configurazione: Development Any CPU ------
3>------ Inizio compilazione: Progetto: EpicGames.OIDC, Configurazione: Development Any CPU ------
4>------ Inizio compilazione: Progetto: EpicGames.Core, Configurazione: Development Any CPU ------
3>EpicGames.OIDC -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.OIDC\bin\Development\net8.0\EpicGames.OIDC.dll
2>EpicGames.Oodle -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.Oodle\bin\Development\net8.0\EpicGames.Oodle.dll
1>EpicGames.UBA -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.UBA\bin\Development\net8.0\EpicGames.UBA.dll
4>EpicGames.Core -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.Core\bin\Development\net8.0\EpicGames.Core.dll
5>------ Inizio compilazione: Progetto: EpicGames.MsBuild, Configurazione: Development Any CPU ------
6>------ Inizio compilazione: Progetto: EpicGames.IoHash, Configurazione: Development Any CPU ------
6>EpicGames.IoHash -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.IoHash\bin\Development\net8.0\EpicGames.IoHash.dll
7>------ Inizio compilazione: Progetto: EpicGames.Serialization, Configurazione: Development Any CPU ------
5>EpicGames.MsBuild -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.MsBuild\bin\Development\net8.0\EpicGames.MsBuild.dll
8>------ Inizio compilazione: Progetto: EpicGames.Build, Configurazione: Development Any CPU ------
7>EpicGames.Serialization -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.Serialization\bin\Development\net8.0\EpicGames.Serialization.dll
9>------ Inizio compilazione: Progetto: EpicGames.Horde, Configurazione: Development Any CPU ------
8>EpicGames.Build -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.Build\bin\Development\net8.0\EpicGames.Build.dll
9>EpicGames.Horde -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.Horde\bin\Development\net8.0\EpicGames.Horde.dll
10>------ Inizio compilazione: Progetto: EpicGames.UHT, Configurazione: Development Any CPU ------
10>EpicGames.UHT -> C:\Program Files\Epic Games\UE_5.5\Engine\Source\Programs\Shared\EpicGames.UHT\bin\Development\net8.0\EpicGames.UHT.dll
11>------ Inizio compilazione: Progetto: UnrealBuildTool, Configurazione: Development Any CPU ------
11>UnrealBuildTool -> C:\Program Files\Epic Games\UE_5.5\Engine\Binaries\DotNET\UnrealBuildTool\UnrealBuildTool.dll
12>------ Inizio compilazione: Progetto: UEReferenceApp, Configurazione: DebugGame_Editor x64 ------
12>Using bundled DotNet SDK version: 8.0.300
12>Running UnrealBuildTool: dotnet "..\..\Engine\Binaries\DotNET\UnrealBuildTool\UnrealBuildTool.dll" UEReferenceAppEditor Win64 DebugGame -Project="E:\Lavori\Unreal\jsbsim5.5\UnrealEngine\UEReferenceApp.uproject" -WaitMutex -FromMsBuild -architecture=x64
12>Log file: C:\Users\adl\AppData\Local\UnrealBuildTool\Log.txt
12>Using 'git status' to determine working set for adaptive non-unity build (E:\Lavori\Unreal\jsbsim5.5).
12>Creating makefile for UEReferenceAppEditor (no existing makefile)
12>Total execution time: 0.38 seconds
12>Unable to find plugin 'Bridge' (referenced via UEReferenceApp.uproject). Install it and try again, or remove it from the required plugin list.
12>C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.MakeFile.Targets(44,5): error MSB3073: uscita dal comando ""C:\Program Files\Epic Games\UE_5.5\Engine\Build\BatchFiles\Build.bat" UEReferenceAppEditor Win64 DebugGame -Project="E:\Lavori\Unreal\jsbsim5.5\UnrealEngine\UEReferenceApp.uproject" -WaitMutex -FromMsBuild -architecture=x64" con codice 6.
12>Compilazione progetto "UEReferenceApp.vcxproj" NON COMPLETATA.
========== Compilazione: 11 completato/i, 1 non riuscito/i, 0 aggiornato/i, 0 ignorato/i ==========
========== Compilazione completato alle 17:40 e ha richiesto 01,689 secondi ==========
